internal/graph/graph.go

```diff
@@ -332,6 +335,61 @@ nextSample:
 	return selectNodesForGraph(nodes, o.DropNegative)
 }
 
+// Trims a Graph that is in forest form to contain only the nodes in kept. This
+// will not work correctly in the case that a node has multiple parents.
+func (g *Graph) TrimTree(kept NodeSet) *Graph {
+	// Creates a new list of nodes
+	oldNodes := g.Nodes
+	g.Nodes = make(Nodes, 0, len(kept.Ptr))
+
+	for _, cur := range oldNodes {
+		// A node may not have multiple parents
+		if len(cur.In) > 1 {
+			fmt.Fprintf(os.Stderr, "ERROR: TrimTree only works on trees.\n")
+		}
+
+		// If a node should be kept, add it to the next list of nodes
+		if _, ok := kept.Ptr[cur]; ok {
+			g.Nodes = append(g.Nodes, cur)
+			continue
+		}
+
+		// Get the parent. Since cur.In may only be of size 0 or 1, parent will be
+		// equal to either nil or the only node in cur.In
+		var parent *Node
+		for _, edge := range cur.In {
+			parent = edge.Src
+		}
+
+		if parent != nil {
```

Comment:
The check could be: if len(cur.In) == 0 {
...

And then you get the parent only after the continue